Revisions (#10)
everywhere-computer/writing-functions-blogpost-2024: BLOGPOST.md

This post was published on the [Fission][fission-blog] blog.

_Authors_: [Brian Ginsburg][bgins] and [Zeeshan Lakhani][zeeshan-lakhani]

---

[Everywhere Computer][everywhere-comp] is an emerging decentralized platform that aims to distribute computational tasks across a vast, open network. This network spans from your personal machine to other devices on your LAN, a cluster of cloud nodes, and even to [PoPs (points of presence)][pop] located at the edge of the Internet. Processing happens as close to the data source as possible or scheduled on nodes with general availability or critical resources like sufficiently capable CPUs or GPUs.

At its core, Everywhere Computer is built on the [InterPlanetary Virtual Machine (IPVM)][ipvm] protocol. It executes [workflows][workflows] containing tasks that are [content-addressed][content-addressing]—which means they're uniquely identified by their content rather than by their location. This system is powered by our [Homestar runtime][homestar-runtime], an engine that runs Wasm-based workflows composed of [Wasm components][wasm-component] with runnable functions that can be scheduled and executed by any Homestar peer throughout the network.

> Beyond the sandboxing, portability, and predictable performance benefits of Wasm, we're excited about orchestrating workflows and state machines composed of modules compiled from different source languages and bringing them together in workflows where the output of one task feeds into the input of another. Composing components in a workflow lets users focus on component interfaces without having to consider interactions between multiple languages.

Reviewer: ```suggestion
Beyond the sandboxing, portability, and predictable performance benefits of Wasm, we're excited about orchestrating workflows and state machines composed of modules compiled from different source languages and bringing them together into workflows where the output of one task feeds into the input of another. Composing components in a workflow lets users focus on component interfaces without considering interactions between multiple languages.
```

![everywhere-at-a-glance](./assets/blogcompute_nontransparent.png)

With Everywhere Computer, we're all in on "[the return of write once, run anywhere][write-once-run]" as a motto, but with content-addressing and our focus on caching and replayability of previously computed tasks, we can go a step further and say "**write once, run once, and never again (everywhere)**."

> With this post, our goal is to introduce authoring Wasm components and, therefore, functions for Everywhere Computer. Wasm components can be authored in [various languages][wit-guest][^1], but we'll focus primarily on Rust, JavaScript, and Python for this post. We'll be writing functions in each of these languages, compiling them to Wasm, packaging them as Wasm components, and bringing them together into a workflow that executes on our compute platform. Along the way, we'll introduce Wasm component tooling, the Homestar runtime, and [Every CLI][everycli], the latter of which provides a convenient interface for running Homestar with a gateway for preparing and executing workflows.

Reviewer: ```suggestion
With this post, our goal is to introduce authoring Wasm components and, therefore, functions for Everywhere Computer. Wasm components can be authored in [various languages][wit-guest][^1], but we'll focus primarily on Rust, JavaScript, and Python for this post. We'll be writing functions in each of these languages, compiling and packaging them as Wasm components, and bringing them together into a workflow that executes on our compute platform. Along the way, we'll introduce Wasm component tooling, the Homestar runtime, and [Every CLI][everycli], the latter of which provides a convenient interface for running Homestar with a gateway for preparing and executing workflows.
```

### Our functions

We will write arithmetic operations in each source language to keep our example code simple. Our Rust program will perform addition and division; the JavaScript one will perform subtraction; and, the Python program will carry out multiplication. We will use division to show division by zero error reporting.

Our functions will be compiled into Wasm components using tools from or built upon the excellent work of the Bytecode Alliance. The Wasm component ecosystem is evolving quickly, so keep in mind that the techniques described in this blog post may be out of date. We'll provide links so you can check on the latest developments.

> Clone the [writing-functions-blogpost-2024][writing-functions-repo] repository if you would like to follow along. You will install tooling for each language, build a component from each language, start Every CLI using the components, and run workflows that call the functions in the components.

Reviewer: ```suggestion
Clone the [writing-functions-blogpost-2024][writing-functions-repo] repository if you would like to follow along. The repository includes instructions for installing dependencies, tooling, and compiling components for each language. We will use Every CLI to run workflows that call the functions in these components.
```

### Workflows

> We now have a set of Wasm components with arithmetic functions sourced from multiple languages. Our next step is to run these functions together in [workflows][workflows].

Reviewer: ```suggestion
We now have a set of Wasm components with arithmetic functions sourced from multiple languages. Let's run these functions together in a [workflow][workflow]!
```
